tinygrad/runtime/ops_metal.py

```diff
@@ -84,15 +92,40 @@ def copyin(self, dest:Any, src:memoryview): self.as_buffer(dest)[:] = src
   def copyout(self, dest:memoryview, src:Any): dest[:] = self.as_buffer(src)
 
 class MetalDevice(Compiled):
+  captureManager = None
+
   def __init__(self, device:str):
     self.device = Metal.MTLCreateSystemDefaultDevice()
+    self.dname = device
+
+    if MTL_CAPTURE_ENABLED and not MetalDevice.captureManager:
+      MetalDevice.captureManager = Metal.MTLCaptureManager.sharedCaptureManager()
+      captureDescriptor = Metal.MTLCaptureDescriptor.alloc().init()
+      captureDescriptor.setDestination_(Metal.MTLCaptureDestination(Metal.MTLCaptureDestinationGPUTraceDocument))
+      captureDescriptor.setOutputURL_(Foundation.NSURL.URLWithString_(f"file:/tmp/tinygrad_{multiprocessing.current_process().name}.gputrace"))
+      captureDescriptor.setCaptureObject_(self.device)
+      unwrap2(MetalDevice.captureManager.startCaptureWithDescriptor_error_(captureDescriptor, None))
+      atexit.register(MetalDevice.stopCapture)
+
     self.mtl_queue = self.device.newCommandQueueWithMaxCommandBufferCount_(1024)
+    self.mtl_queue.setLabel_(self.dname)
     self.mtl_buffers_in_flight: List[Any] = []
     self.mv_in_metal: List[memoryview] = []
     from tinygrad.runtime.graph.metal import MetalGraph
     super().__init__(device, MetalAllocator(self), MetalCompiler(None if getenv("METAL_XCODE") else self),
                      functools.partial(MetalProgram, self), functools.partial(MetalGraph, self))
+
+  @staticmethod
+  def stopCapture():
+    if MetalDevice.captureManager:
+      MetalDevice.captureManager.stopCapture()
+      print(f"Saved GPU trace to /tmp/tinygrad_{multiprocessing.current_process().name}.gputrace")
+
   def synchronize(self):
-    for cbuf in self.mtl_buffers_in_flight: wait_check(cbuf)
+    if len(self.mtl_buffers_in_flight) > 0:
+      self.mtl_buffers_in_flight[-1].waitUntilCompleted()
```

Comment:
Err, is this okay to assume they are in order?

Reply:
They're added to MTLCommandQueue and this list in the same order and should be executed in same order

I did it to declutter list of API calls, can revert this change, it isn't very important